spec.bs

```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+  1. If |regexp or wildcard token| is null, then set |regexp value| to |parser|'s [=pattern parser/segment wildcard regexp=].
+  1. Else if |regexp or wildcard token|'s [=token/type=] is "<a for=token/type>`asterisk`</a>", then set |regexp value| to the [=full wildcard regexp value=].
+  1. Else set |regexp value| to |regexp or wildcard token|'s [=token/value=].
+  1. Let |type| be "<a for=part/type>`regexp`</a>".
+  1. If |regexp value| is |parser|'s [=pattern parser/segment wildcard regexp=]:
+    1. Set |type| to "<a for=part/type>`segment-wildcard`</a>".
+    1. Set |regexp value| to the empty string.
+  1. Else if |regexp value| is the [=full wildcard regexp value=]:
+    1. Set |type| to "<a for=part/type>`full-wildcard`</a>".
+    1. Set |regexp value| to the empty string.
+  1. Let |name| be the empty string.
+  1. If |name token| is not null, then set |name| to |name token|'s [=token/value=].
+  1. Else if |regexp or wildcard token| is not null:
+    1. Set |name| to |parser|'s [=pattern parser/next numeric name=].
+    1. Increment |parser|'s [=pattern parser/next numeric name=] by 1.
+  1. Let |encoded prefix| be the result of running |parser|'s [=pattern parser/encoding callback=] given |prefix|.
+  1. Let |encoded suffix| be the result of running |parser|'s [=pattern parser/encoding callback=] given |suffix|.
+  1. Set |part|'s [=part/type=] to |type|.
+  1. Set |part|'s [=part/name=] to |name|.
+  1. Set |part|'s [=part/prefix=] to |encoded prefix|.
+  1. Set |part|'s [=part/value=] to |regexp value|.
+  1. Set |part|'s [=part/suffix=] to |encoded suffix|.
+  1. Set |part|'s [=part/modifier=] to |modifier|.
+  1. [=list/Append=] |part| to |parser|'s [=pattern parser/part list=].
+</div>
+
+<div algorithm>
+To <dfn export>parse a pattern</dfn> given a [=/pattern string=] |input|, [=/options=] |options|, and [=/encoding callback=] |encoding callback|:
+
+  1. Let |parser| be a new [=pattern parser=].
+  1. Set |parser|'s [=pattern parser/options=] to |options|.
+  1. Set |parser|'s [=pattern parser/encoding callback=] to |encoding callback|.
+  1. Set |parser|'s [=pattern parser/token list=] to the result of running [=tokenize=] given |input| and "<a for="tokenize policy">`strict`</a>".
+  1. While |parser|'s [=pattern parser/index=] is less than |parser|'s [=pattern parser/token list=]'s [=list/size=]:
+    1. Let |char token| be the result of running [=try to consume a token=] given |parser| and "<a for=token/type>`char`</a>".
+    1. Let |name token| be the result of running [=try to consume a token=] given |parser| and "<a for=token/type>`name`</a>".
+    1. Let |regexp or wildcard token| be the result of running [=try to consume a token=] given |parser| and "<a for=token/type>`regexp`</a>".
+    1. If |name token| is null and |regexp or wildcard token| is null:
+      1. Set |regexp or wildcard token| to the result of running [=try to consume a token=] given |parser| and "<a for=token/type>`asterisk`</a>".
```

Comment:
Probably this doesn't need to be nested.

Reply:
Done.  And I factored it out to a separate algorithm since we need to do these steps at least twice.